src/rust/engine/engine_cffi/src/lib.rs

```diff
@@ -889,20 +893,45 @@ pub extern "C" fn materialize_directories(
   };
   let workunit_store = with_session(session_ptr, |session| session.workunit_store());
   with_scheduler(scheduler_ptr, |scheduler| {
-    scheduler.core.executor.block_on(
-      futures::future::join_all(
-        dir_and_digests
-          .into_iter()
-          .map(|(dir, digest)| {
-            scheduler
-              .core
-              .store()
-              .materialize_directory(dir, digest, workunit_store.clone())
-          })
-          .collect::<Vec<_>>(),
-      )
-      .map(|_| ()),
+    let types = &scheduler.core.types;
+    let construct_materialize_directories_results = types.construct_materialize_directories_results;
+    let construct_materialize_directory_result = types.construct_materialize_directory_result;
+    let work_future = futures::future::join_all(
+      dir_and_digests
+        .into_iter()
+        .map(|(dir, digest)| {
+          scheduler
+            .core
+            .store()
+            .materialize_directory(dir, digest, workunit_store.clone())
+        })
+        .collect::<Vec<_>>(),
     )
+    .map(move |metadata_list| {
+      let entries: Vec<Value> = metadata_list
+        .iter()
+        .map(|metadata: &store::DirectoryMaterializeMetadata| {
+          let path_list = metadata.to_path_list();
+          let path_values: Vec<Value> = path_list
+            .into_iter()
+            .map(|path: String| externs::store_utf8(&path))
+            .collect();
+
+          externs::unsafe_call(
+            &construct_materialize_directory_result,
+            &[externs::store_tuple(&path_values)],
+          )
+        })
+        .collect();
+
+      let output: Value = externs::unsafe_call(
+        &construct_materialize_directories_results,
+        &[externs::store_tuple(&entries)],
+      );
+      output
+    });
+
+    scheduler.core.executor.spawn_on_io_pool(work_future).wait()
```

Comment:
The need to use the IO pool like this only applies to this usecase, so it might be good to have two versions of this method? On the other hand, this method has always been blocking, and this is just a different way to block.

I'm fine with leaving this, but please add a big comment that explains that we do actually call this method from the tokio runtime in some cases, and that that will block a runtime thread. It would probably also be good to ensure that we always have at least one thread.